commons/src/main/java/org/infinispan/commons/util/IntSets.java

```diff
@@ -0,0 +1,192 @@
+      if (integerSet instanceof SmallIntSet) {
+         return (SmallIntSet) integerSet;
+      }
+      if (integerSet instanceof ConcurrentSmallIntSet) {
+         return (ConcurrentSmallIntSet) integerSet;
+      }
+      return mutableCopyFrom(integerSet);
+   }
+
+   /**
+    * Returns an IntSet that contains all ints from the given Set that is mutable. Updates to the original Set or
+    * the returned IntSet are not reflected in the other.
+    * @param mutableSet set to copy from
+    * @return IntSet with the values set
+    */
+   public static IntSet mutableCopyFrom(Set<Integer> mutableSet) {
+      if (mutableSet instanceof SingletonIntSet) {
+         return mutableSet(((SingletonIntSet) mutableSet).value);
+      }
+      return new SmallIntSet(mutableSet);
+   }
+
+   /**
+    * Returns an IntSet that contains no values but is initialized to hold ints equal to the {@code maxExclusive -1} or
+    * smaller.
+    * @param maxExclusive largest int expected in set
+    * @return IntSet with no values set
+    */
+   public static IntSet mutableEmptySet(int maxExclusive) {
+      return new SmallIntSet(maxExclusive);
+   }
+
+   /**
+    * Returns a mutable IntSet with no values set. Note this mutable set is initialized given a default size. If you wish
+    * to not have less initialization over, please use {@link #mutableEmptySet(int)} providing a {@code 0} or similar.
+    * @return IntSet with no values set
+    */
+   public static IntSet mutableEmptySet() {
+      return new SmallIntSet();
+   }
+
+   /**
+    * Returns a mutable set with the initial value set. This set is optimized to insert values less than this. Values
+    * added that are larger may require additional operations.
+    * @param value the value to set
+    * @return IntSet with the value set
+    */
+   public static IntSet mutableSet(int value) {
+      return SmallIntSet.of(value);
+   }
+
+   /**
+    * Returns a mutable IntSet that begins with the initialized values
+    * @param value1
+    * @param value2
+    * @return
+    */
+   public static IntSet mutableSet(int value1, int value2) {
+      return SmallIntSet.of(value1, value2);
+   }
+
+   /**
+    * Returns a concurrent mutable IntSet that can store values in the range of {@code 0..maxExclusive-1}
+    * @param maxExclusive the maximum value - 1 that can be inserted into the set
+    * @return concurrent set
+    */
+   public static IntSet concurrentSet(int maxExclusive) {
+      return new ConcurrentSmallIntSet(maxExclusive);
+   }
+
+   /**
+    * Returns a copy of the given set that supports concurrent operations. The returned set will contain all of the
+    * ints the provided set contained. The returned set only supports up to the maximum size the previous int set
+    * supported when this method is invoked or the largest int it held.
+    * @param intSet set to copy from
+    * @return concurrent copy
+    */
+   public static IntSet concurrentCopyFrom(IntSet intSet) {
```

Comment:
In general I'm not convinced we need to hide constructors between static methods. But especially here I would prefer a constructor `ConcurrentSmallIntSet(intSet, maxSize)` that passes the capacity explicitly instead of guessing it from the set parameter.

Reply:
Sounds good.